### ISSUE TITLE
Object.defineProperty getter and setter does not work

### ISSUE THREAD

**Issue by Evol-Intent** [cesanta/v7#507]
The code is:
```javascript
(function(global) {

	var Player = function(num) {
		this.num = num;
	};

	Object.defineProperty(Player.prototype, 'number', {
		get: function() {
			return this.num+1;
		},
		set: function(value) {
			this.num = value*10;
		}
	});

	var p = new Player(7);
	print(p.number);
	p.number=2;
	print(p.number);

})(this);
```

The result should be: 8, 21
But the result is: undefined, undefined

**Comment by mmikulicic** [comment]
Yes, in order to simplify the VM internals and make it fit in the smallest microcontrollers we decided to skip the implementation of getters and setters for now. Only system C functions can be setters and getters (such as Array.prototype.length)

We might lift this restriction in the future.

**Comment by mmikulicic** [comment]
We should throw an exception though I guess

**Comment by Evol-Intent** [comment]
Thank you for the quick response.

As long as this feature is not supported, is it possible to achieve its effect somehow without a deep knowledge about the v7 engine?

**Comment by mmikulicic** [comment]
Yes, but you have to write the getter and setter functions in C. 
You don't have to have deep knowledge about the V7 engine, but there are a few things to mention

1. This are is currently severely underdocumented.
2. The setter and getter function must not reenter the interpreter (e.g. by calling a JS function or evaluating code)
3. The setter and getter run with GC inhibited and you must not disinhibit it.
4. The API to set both getter and setter is clumsy and reveals internals.

```c
v7_val_t getter = v7_create_function(v7, my_getter, 0);
v7_val_t setter = v7_create_function(v7, my_setter, 1);
v7_val_t getset = v7_create_array(v7);
v7_array_set(v7, my_getter, 0);
v7_array_set(v7, my_setter, 1);
v7_set_property(v7, obj, "number", ~0, V7_PROPERTY_DONT_ENUM | V7_PROPERTY_SETTER | V7_PROPERTY_GETTER,
                         getset);
```

The array is needed to hold both setter and getter under the same slot. We might come up with a better solution or at least a better API; sorry about that.

Note: You might have noticed that V7 supports JS getter and setters defined in during object creation: `o={get a() { return 12 }}; o.a`. This feature is going away in order to simplify internals; we might bring it back. We want to focus on core feature set first. We will keep the aforementioned C interface though.

**Comment by Evol-Intent** [comment]
Thank you, I will check it out.

If you remove the ability to define getter and setter during object creation, or you consider later to implement Object.defineproperty function, would you like to enable it under a build option? Like -DV7_GETTER_SETTER

So v7 could still fit in the smallest microcontrollers, but the feature could be enabled for those who need it.

**Comment by mmikulicic** [comment]
Yes, it will be an optional feature enabled with a build option.

Just for the record: the complicated thing is not the getter/setters per se, but the structures in the interpreter required to cope with the GC being triggered potentially every time someone calls `v7_get` without OOMing on heaps of about 10KB. This constrain will be relaxed in the next few rounds of improvements of the VM, but we cannot give an ETA at the moment.

**Comment by HaydnTrigg** [comment]
Are there any recent updates on this feature?

**Comment by payalpatel26** [comment]
var Player = function(num) {
        this.num = num;
    };

    Object.defineProperty(Player.prototype, 'number', {
        get: function() {
            return this.num+1;
        },
        set: function(value) {
            this.num = value*10;
        }
    });

    var p = new Player(7);
    print(p.number);
    p.number=2;
    print(p.number);
if you run only this code you will get 8 and 21 i think the problem in function(global)
